ale_linters/go/gometalinter.vim

```diff
@@ -26,10 +34,12 @@ function! ale_linters#go#gometalinter#GetMatches(lines) abort
 endfunction
 
 function! ale_linters#go#gometalinter#Handler(buffer, lines) abort
+    let l:dir = expand('#' . a:buffer . ':p:h')
     let l:output = []
 
     for l:match in ale_linters#go#gometalinter#GetMatches(a:lines)
         call add(l:output, {
+        \   'filename': ale#path#GetAbsPath(l:dir, l:match[1]),
```

Comment:
The changes to this function not only cause much longer filenames, but is also causes my statusline to stop showing me the errors/warnings counts and it stops the error and warning signs to show up in the signcolumn.

So this part actually causes 3 different (yet maybe related) issues that should be fixed before we merge this PR.

Reply:
I agree this part might need changing, but I have too many questions about how the quickfix window handles the absolute filenames. 

Something changes in the way they are interpreted when there is a buffer change. I don't know enough of the internals to know what is happening, but I have seen it change from an absolute filename to a "relative to the current buffer" filename when a lint even happens and then the buffer is changed. 

It is possible that ale needs to change somewhere else to only display the relative path to the user in the quickfix window even though ale itself has received the absolute path. It is also possible that the linter needs to give a relative path with something like `ale#path#GetRelPath` I am not sure, @w0rp do you have any input on how to handle this?

Reply:
I opened an issue for doing something about long filenames here https://github.com/w0rp/ale/issues/1160

You must use absolute paths for the filename key to know which files errors are for, for other files.

@svanharmelen What is your actual problem? Is the problem that you aren't seeing errors in a file where you did before? Maybe we could just not include the `filename` key if the path is the same as the buffer's path.

Reply:
@w0rp well, it seems that using this code gives me 3 problems:

1. Long filenames that fill up the loclist, leaving little room for the actual error/warning. So this means every finding uses multiple lines which makes it very annoying to read;
2. I don't get the error/warning count in the statusline anymore;
3. I don't get the error/warnings signs in the signcolumn.

Reply:
It sounds like the filename key is just wrong then. If the filename matched the buffer's filename, you'd see the errors and warnings still. That means that the paths are relative to some other directory than the current file's directory. To fix that, you'll have to figure out which directory the paths are actually relative to. You might have to prefix the command with `cd file_directory &&` if the paths are just relative to whatever the working directory is.

Reply:
just to clarify, it sounds like you are saying that ale is working as expected and that the clash is coming from somewhere else in the `vimrc`?

@svanharmelen can you checkout this commit and see if your problem persists even with the `gometalinter_lint_package` setting in its default state of `0`?

Reply:
I'm saying the filename isn't correct on @svanharmelen's machine. If he sees valid errors for a file, uses this patch, and the errors disappear, then the filename has to be wrong. That means that the paths from the output aren't relative to the directory we are using.

Reply:
What I did earlier today is checkout this commit and test if things kept working as expected. So I didn't change my `init.vim` (I'm using NeoVim) as the default for `go_gometalinter_lint_package` is already set to `0` and so it should not change anything, right?

As for the problem(s) we are discussing, I do use [vim-rooter](https://github.com/airblade/vim-rooter) to always set my `pwd` to the project root. So I guess that could cause the differences in paths you are refering to @w0rp?

Reply:
You might need to prefix the command with `ale#path#BufferCdString(a:buffer)` to get it to work everywhere, which will make the working directory for the command predictable. You might need to do something else.

Reply:
Just out of curiosity... What are we trying to fix or achieve by adding the line:

```
  \ 'filename': ale#path#GetAbsPath(l:dir, l:match[1]),
```

to the output? As apparently I'm missing a use case and/or a purpose of why we are trying to change this... 🤓

Reply:
It's for showing errors for _other files_.

If you have a file `foo.go` which imports `bar.go`, you can show errors for `foo.go` and also errors in the imported `bar.go` file, which might be the cause of problems. As long as the correct filenames are used, the only difference you should notice is that some more information appears in the quickfix window or the loclist window.